tsdb/head.go

```diff
@@ -213,6 +213,32 @@ func NewHead(r prometheus.Registerer, l log.Logger, wal *wal.WAL, opts *HeadOpti
 	return h, nil
 }
 
+func (h *Head) CheckOutOfOrderSample(ref uint64, t int64, v float64) bool {
+	s := h.series.getByID(ref)
+	if s == nil || s.head() == nil {
+		return false
+	}
+
+	previousMatched := false
+	switch {
+	case math.Float64bits(s.sampleBuf[3].v) == math.Float64bits(v) && s.sampleBuf[3].t == t:
+		previousMatched = true
+	case math.Float64bits(s.sampleBuf[2].v) == math.Float64bits(v) && s.sampleBuf[2].t == t:
+		previousMatched = true
+	case math.Float64bits(s.sampleBuf[1].v) == math.Float64bits(v) && s.sampleBuf[1].t == t:
+		previousMatched = true
+	case math.Float64bits(s.sampleBuf[0].v) == math.Float64bits(v) && s.sampleBuf[0].t == t:
+		previousMatched = true
+	}
+
+	if !previousMatched && t < s.head().maxTime {
+		h.metrics.outOfOrderSamples.Inc()
```

Comment:
This function is just to get some info; the sample might not be attempted to be added later. So we should not update any metric in this function.